Fix new warning messages with Elixir 1.3

jfrolich opened the pull request:
Warnings stopped working for me with linter-elixirc. With the following change I got it back working

Arcanemagus commented:
This will need to be backwards compatible if at all possible.

Is there any way to customize (stabilize) the output so we don't keep running into this issue?

jfrolich commented:
Hey @Arcanemagus. Good point. I removed the other regex, but this will indeed break backwards compatibility. Will update this. I am testing a new version right now.

 Any specific reason why you would want to compile with mix instead of elixirc? Was just wondering while working on the source. Mix has a few downsides in that it does not show the warnings on a second recompile, so I was thinking of removing compile with mix support.

jfrolich commented:
By the way, I do not think there is a way to stabilize this if there will be another change to the compiler's output. Unless there will be another way of communicating with the compiler instead of text output.

Arcanemagus commented:
You'd have to ask @AtomLinter/linter-elixirc about the `compile` questions, I don't use this linter myself.

Stabilizing the output would depend on `elixirc` supporting an option to specify it, as I believe this is actually a compiler that we are just parsing the output of, this is probably unlikely. That was a question instead of a "you need to do this" 😛.

jfrolich commented:
@Arcanemagus: that link does not work. You are one of the authors of this package right? Anyway I do not really see a use of linting with `mix compile` instead of `elixirc` (to codepaths in this source) so I plan to strip it out in this pull request, but it might serve some purpose?

remiq commented:
@jfrolich: `elixirc` used to compile only current file and `mix compile` allowed to lint whole project. Not sure how it is now.

Arcanemagus commented:
@jfrolich I maintain the entire AtomLinter organization, and have made some fixes here, but I don't use this linter or really know anything about it. I really wish the teams could be publicly visible so outside people could see the list of maintainers for a project 😞.

jfrolich commented:
Alright this should fix legacy support.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -103,7 +103,41 @@ module.exports =
+            type: "Warning"
+            text: reResult[1]
+            filePath: path.join(projectPath(textEditor), reResult[2])
+            # use range, this because the previous method of getting a range
```
Instead of completely abandoning the nice ranges provided by `rangeFromLineNumber`, how about we parse through [`atom.workspace.getTextEditors()`](https://atom.io/docs/api/v1.10.2/Workspace#instance-getTextEditors), checking the path they provide and send the proper editor to that if one is available?

If one isn't available, then a range like this is fine as it will be updated when that file is worked on, however the range should be a character wide, otherwise no highlight will be shown, and the cursor will be hidden if it is clicked on:

```coffeescript
new Range(
  [reResult[3] - 1, 0],
  [reResult[3] - 1, 1]
)
```

Arcanemagus commented on lib/init.coffee:
```diff
@@ -103,7 +103,41 @@ module.exports =
+      reResult = re.exec(toParse)
+
+      while reResult?
+        console.log(reResult)
```
This should be wrapped in a check on `atom.inDevMode()`, if not removed entirely.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -113,22 +147,31 @@ module.exports =
           errorStack = parseError(resultString, textEditor)
           warningStack = parseWarning(resultString, textEditor)
+          legacyWarningStack = parseLegacyWarning(resultString, textEditor)
           (error for error in errorStack.concat(warningStack) when error?)
```
`legacyWarningStack` isn't being concatenated into the `error` list.

Arcanemagus commented on lib/init.coffee:
```diff
@@ -113,22 +147,31 @@ module.exports =
+          legacyWarningStack = parseLegacyWarning(resultString, textEditor)
           (error for error in errorStack.concat(warningStack) when error?)
         catch Error
+          console.log(Error, "oops caught error....")
```
Prefix this with an identifier as to where it is coming from (`linter-elixirc: `).

Arcanemagus commented on lib/init.coffee:
```diff
@@ -103,7 +103,41 @@ module.exports =
+            # if the compiler returned warnings or errors of other files,
+            # it would thus frequently blow up and return no errors/warnings
+            # at all
+            range: new Range(
```
We don't _need_ to bring in Range here btw, a "Range compatible array" works just fine. It doesn't hurt anything to do it this way though :wink:.

Arcanemagus commented:
Sorry it took so long to review the changes, was sick most of the weekend :mask:.

jfrolich commented:
Sorry for my (very) late response as well. I converted this package to es2015 as I'm not as proficient in coffeescript anymore, and it is the new standard for atom. Hope you agree with this.
I also addressed your comments.

Arcanemagus commented:
Hmmmm, do you want a detailed review of the code? I'm torn between just making sure this works and then doing the rest of the rewrite, or doing a _massive_ review here to bring this in line with all the changes I normally incorporate when updating a project away from that horrid mess that is CoffeeScript.

It's completely up to you, I just don't want the work you put in here to be completely lost.

jfrolich commented:
Both are fine with me. Maybe updating the code is less work for you than a review?

Arcanemagus commented:
Fixing it up later is certainly easier for me 😛.

As it doesn't look like this project even has CI setup right now I'll just merge this as is and then do a cleanup PR, thanks for getting this started!

jfrolich commented:
Functionally it works great for me right now on my projects. Looking forward to your changes. This is the first pull-request to an atom package, so probably much to improve :)

jfrolich commented:
By the way, I used prettier for es2015 formatting (using prettier-atom). It works great, and applies a consistent code style.

Arcanemagus commented:
@jfrolich If you were curious, pushed up the likely functional changes so far to [arcanemagus/cleanup](https://github.com/AtomLinter/linter-elixirc/tree/arcanemagus/cleanup), still need to write specs, validate that it actually works, and finish any more refactoring that comes up while working on that.
